### DISCUSSION TITLE
environment variables in baseUrl?

### DISCUSSION THREAD

**Discussion by marc-guenther** [cypress-io/cypress#9321] [Questions and Help]
Ok, this is rather strange and also somewhat silly, lemme try to explain...

We run Cypress on our CI system. The application it tests is build and deployed inside a Vagrant box, and the box forwards its port 80 to some port on the CI machine. As multiple test jobs can run at the same time, we **cannot use a fixed port** number for this, so instead we allocate a random port, and expose it as an environment variable: `VAGRANT_HTTP_PORT`

To make this **port** available to all tests, I put this in my `cypress.json`:
```
{
  "baseUrl": "http://localhost:${VAGRANT_HTTP_PORT}/",
}
```

Unfortunately Cypress doesn't support env variables in the `baseUrl`, so I had to do it manually in `cypress/support/index.js`:
```
// Replace env variables (eg. VAGRANT_HTTP_PORT) in the baseUrl from cypress.json:
Cypress.config("baseUrl", Cypress.config().baseUrl.replace(/\$\{(\w+)\}/g, (s,g1) => Cypress.env(g1)||s))
```

That wasn't too hard, and it **works perfectly fine** on my Macbook, and all the CI Macs, all running Catalina. **Except** on one machine.

On the laptop of a collegue, when starting Cypress, she gets an error message, that says something like: **`http://localhost:${VAGRANT_HTTP_PORT}/` cannot be reached**. For some reason, the env variable is **not replaced** there, although it definitely has been set.

Of course, this happens on the **only** machine, that, due to current lockdown, I do not really have access to. Which makes it rather hard to debug the problem. It is also running Catalina, it is the **same** Cypress version, running from the **same** repository. I am not aware of any significant difference in her machine vs. all the others.

Our current workaround is for her to hardcode the port in the `baseUrl`, and to remember to not check in that change. Not ideal, but it works.

Any ideas are highly appreciated.

**Comment by bahmutov** [comment]
Look at the Settings tab in `cypress open` - it should show every variable and where it comes from

<img width="800" alt="Screen Shot 2020-11-24 at 5 53 55 PM" src="https://user-images.githubusercontent.com/2212006/100160758-2f583900-2e7e-11eb-9fc6-2bf5847351d6.png">

Second, do not set the base URL in the support file based on the environment variables, do it in the plugins file https://on.cypress.io/configuration-api (and you can even print the final variables to the terminal)

**Reply by marc-guenther** [reply]
Ok, this is interesting. On her machine, she get's the warning message, and the entry for `baseUrl` is in blue (config): `baseUrl:"http://localhost:${VAGRANT_HTTP_PORT}"`

On my machine, I get the *same* entry in blue, but NO warning message about that URL being unreachable.

Also, on both of our machines, it actually works if we just click "Run all specs". So it seems, the variable is actually being replaced.

Whereas, running from command line `cypress run`, it only works on my machine and on her machine it again fails with the error that it cannot reach the URL `"http://localhost:${VAGRANT_HTTP_PORT}"`.

So, two results:
- I had put the logic into the wrong place. It needs to go into the plugins file. Thanks for pointing that out, I will try it later.
- For some reason, the test which is built into Cypress which checks for a valid baseUrl on Cypress startup, is never executed on my machine or any of the CI machines. In fact, I can put completely nonsense into the baseUrl, and it never complains, whereas on her machine, the URL has to be reachable on startup.
